rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll

```diff
@@ -601,71 +602,134 @@ module LocalFlow {
   }
 }
 
-private import codeql.util.Option
+/**
+ * Provides temporary modeling of built-in variants, for which no source code
+ * `Item`s are available.
+ *
+ * TODO: Remove once library code is extracted.
+ */
+private module VariantLib {
+  private import codeql.util.Option
 
-private class CrateOrigin extends string {
-  CrateOrigin() {
-    this = [any(Item i).getCrateOrigin(), any(Resolvable r).getResolvedCrateOrigin()]
+  private class CrateOrigin extends string {
+    CrateOrigin() { this = any(Resolvable r).getResolvedCrateOrigin() }
   }
-}
 
-private class CrateOriginOption = Option<CrateOrigin>::Option;
+  private class CrateOriginOption = Option<CrateOrigin>::Option;
 
-pragma[nomagic]
-private predicate hasExtendedCanonicalPath(Item i, CrateOriginOption crate, string path) {
-  path = i.getExtendedCanonicalPath() and
-  (
-    crate.asSome() = i.getCrateOrigin()
-    or
-    crate.isNone() and
-    not i.hasCrateOrigin()
-  )
-}
+  private CrateOriginOption langCoreCrate() { result.asSome() = "lang:core" }
 
-pragma[nomagic]
-private predicate variantHasExtendedCanonicalPath(
-  Enum e, Variant v, CrateOriginOption crate, string path, string name
-) {
-  hasExtendedCanonicalPath(e, crate, path) and
-  v = e.getVariantList().getAVariant() and
-  name = v.getName().getText()
-}
+  private newtype TVariantLib =
+    MkVariantLib(CrateOriginOption crate, string path, string name) {
+      crate = langCoreCrate() and
+      (
+        path = "crate::option::Option" and
+        name = "Some"
+        or
+        path = "crate::result::Result" and
+        name = ["Ok", "Err"]
+      )
+    }
 
-pragma[nomagic]
-private predicate resolveExtendedCanonicalPath(Resolvable r, CrateOriginOption crate, string path) {
-  path = r.getResolvedPath() and
-  (
-    crate.asSome() = r.getResolvedCrateOrigin()
+  /** A canonical path pointing to an enum variant. */
+  class VariantLib extends MkVariantLib {
```

Comment:
What about using `VariantInLib` or `VariantFromLib` or something of that sort? My first reading of this was "a QL library for variants" 😅

Reply:
Could also be `VariantInStdLib` assuming we don't extend to take this special casing further than that.